core/src/main/java/io/questdb/cairo/GenericRecordMetadata.java

```diff
@@ -38,39 +38,55 @@ public GenericRecordMetadata() {
         this.timestampIndex = -1;
     }
 
-    public static void copyColumns(RecordMetadata from, GenericRecordMetadata to) {
-        if (from instanceof BaseRecordMetadata) {
-            final BaseRecordMetadata gm = (BaseRecordMetadata) from;
-            for (int i = 0, n = gm.getColumnCount(); i < n; i++) {
-                to.add(gm.getColumnQuick(i));
-            }
-        } else {
-            for (int i = 0, n = from.getColumnCount(); i < n; i++) {
-                to.add(new TableColumnMetadata(
-                        from.getColumnName(i),
-                        from.getColumnHash(i),
-                        from.getColumnType(i),
-                        from.isColumnIndexed(i),
-                        from.getIndexValueBlockCapacity(i),
-                        from.isSymbolTableStatic(i),
-                        GenericRecordMetadata.copyOf(from.getMetadata(i))
-                ));
-            }
-        }
-    }
-
     public static GenericRecordMetadata copyOf(RecordMetadata that) {
         if (that != null) {
             if (that instanceof GenericRecordMetadata) {
                 return (GenericRecordMetadata) that;
             }
-            GenericRecordMetadata metadata = copyOfSansTimestamp(that);
-            metadata.setTimestampIndex(that.getTimestampIndex());
+            GenericRecordMetadata metadata = new GenericRecordMetadata();
+            metadata.setTimestampIndex(copyColumns(that, metadata));
             return metadata;
         }
         return null;
     }
 
+    private static int copyColumns(RecordMetadata from, GenericRecordMetadata to) {
+        int timestampIndex = -1;
+        int fromTimestampIndex = from.getTimestampIndex();
+        if (from instanceof BaseRecordMetadata) {
+            final BaseRecordMetadata gm = (BaseRecordMetadata) from;
+            for (int i = 0, n = gm.getColumnCount(); i < n; i++) {
+                TableColumnMetadata column = gm.getColumnQuick(i);
+                if (column.getType() >= 0) {
+                    to.add(column);
+                    if (i == fromTimestampIndex) {
+                        timestampIndex = to.getColumnCount() - 1;
+                    }
+                }
+            }
+        } else {
+            for (int i = 0, n = from.getColumnCount(); i < n; i++) {
```

Comment:
this branch is not tested and cannot be because currently all implementations of RecordMetadata extends BaseRecordMetadata.
We could get rid of this branch easily by adding this method to RecordMetadata:
```
TableColumnMetadata getColumnQuick(int index)
```
Or just change this method's signature to expect BaseRecordMetadata.